src/IssueViz.Security/Hotspots/LocalHotspotsStore.cs

```diff
@@ -0,0 +1,214 @@
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal sealed class LocalHotspotsStore : ILocalHotspotsStore
+    {
+        private static readonly List<IAnalysisIssueVisualization> EmptyList = new List<IAnalysisIssueVisualization>();
+
+        private readonly object lockObject = new object();
+        private readonly IHotspotMatcher hotspotMatcher;
+        private readonly IThreadHandling threadHandling;
+        private readonly IServerHotspotStore serverHotspotStore;
+        private Dictionary<string, List<LocalHotspot>> fileToHotspotsMapping =
+            new Dictionary<string, List<LocalHotspot>>();
+
+        private List<SonarQubeHotspot> serverHotspots = new List<SonarQubeHotspot>();
+        private HashSet<SonarQubeHotspot> unmatchedHotspots = new HashSet<SonarQubeHotspot>();
+
+        [ImportingConstructor]
+        public LocalHotspotsStore(IServerHotspotStore serverHotspotStore, IHotspotMatcher hotspotMatcher, IThreadHandling threadHandling)
+        {
+            this.serverHotspotStore = serverHotspotStore;
+            this.hotspotMatcher = hotspotMatcher;
+            this.threadHandling = threadHandling;
+            InitializeServerHotspots(this.serverHotspotStore.GetAll());
+            serverHotspotStore.Refreshed += (sender, args) => RematchAll();
+        }
+
+        public event EventHandler<IssuesChangedEventArgs> IssuesChanged;
+
+        public IReadOnlyCollection<IAnalysisIssueVisualization> GetAll()
+        {
+            threadHandling.ThrowIfOnUIThread();
+            
+            lock (lockObject)
+            {
+                return FlattenMapping().Select(x => x.Visualization).ToList();
+            }
+        }
+
+        public List<LocalHotspot> GetAllLocalHotspots()
+        {
+            threadHandling.ThrowIfOnUIThread();
+
+            lock (lockObject)
+            {
+                return FlattenMapping().ToList();
+            }
+        }
+
+        public void AddOrUpdate(string filePath, IEnumerable<IAnalysisIssueVisualization> hotspots)
+        {
+            threadHandling.ThrowIfOnUIThread();
+
+            lock (lockObject)
+            {
+                List<IAnalysisIssueVisualization> oldIssueVisualizations;
+                if (fileToHotspotsMapping.TryGetValue(filePath, out var oldHotspots))
+                {
+                    // possible optimization point. it's unlikely we need to re-match every single hotspot
+                    UnmatchServerHotspots(oldHotspots);
+                    oldIssueVisualizations = oldHotspots
+                        .Select(x => x.Visualization)
+                        .ToList();
+                }
+                else
+                {
+                    oldIssueVisualizations = EmptyList;
+                }
+
+                var hotspotsList = hotspots.ToList();
+                
+                if (!hotspotsList.Any() && !oldIssueVisualizations.Any())
+                {
+                    return;
```

Comment:
`Any` should work efficiently with lazy enumerations, so you could check `hotspots` before making a copy of the list i.e. avoid making a copy unless you need to.